### ISSUE TITLE
Spotify: PulseAudio: Unable to connect: Connection refused

### ISSUE THREAD

**Issue by xexpanderx** [flatpak/flatpak#2169]
## Steps to reproduce

Before starting spotify I issue this in shell:
`
export `xprop -root -notype PULSE_SERVER | tr -d ' '``

and then

`flatpak run com.spotify.Client/x86_64/stable`

which gives me this:
```

ALSA lib ../../pulse/pulse.c:243:(pulse_connect) PulseAudio: Unable to connect: Connection refused

ALSA lib ../../src/dlmisc.c:170:(snd_dlsym_verify) unable to verify version for symbol _snd_pcm_plug_open
ALSA lib ../../src/dlmisc.c:296:(snd1_dlobj_cache_get) symbol _snd_pcm_plug_open is not defined inside [builtin]
```
No idea how to fix this. I have Spotify opened, but no sound.

**Comment by TingPing** [comment]
Does this happen with other Flatpaks?

**Comment by xexpanderx** [comment]
This is the first time I try to use Flatpak, do you have any other package I may try that requires pulseaudio?

**Comment by xexpanderx** [comment]
Hi,

The value is:
```
echo $PULSE_SERVER 
"{ca2c3d36d123493d9a89cae35b7c676a}unix:/tmp/pulse-cKwRHHS8pmwY/native"
```

**Comment by CrashyBang** [comment]
@xexpanderx did you find a solution? I am experiencing the same problem.

**Comment by xexpanderx** [comment]
> @xexpanderx did you find a solution? I am experiencing the same problem.

No, I gave up, too much of a hassle.

**Comment by CrashyBang** [comment]
Fair enough I will comment here if I find anything simple.

**Comment by kirbyfan64** [comment]
Based on #878, this seems to occur if `$XDG_RUNTIME_DIR` is empty when PulseAudio was started. Try adding:

```
export XDG_RUNTIME_DIR=/run/user/$(id -u)
```

**Comment by CrashyBang** [comment]
Hey @kirbyfan64,

I added that to my `.zshrc` and now I get the following error:
```
error: Unable to allocate instance id
```
I also tried install [`xdg-utils`](https://wiki.voidlinux.eu/AppImages) in case this was required but it made no difference.

**Comment by CrashyBang** [comment]
After reading a little more I tried adding:
```
export XDG_RUNTIME_DIR=/run/user/$(id -u)
```
To my `/etc/profile` this stopped the above error, but did not resolve the issues.

**Comment by kirbyfan64** [comment]
Now, what's the value of `$PULSE_SERVER`?

**Comment by CrashyBang** [comment]
Hey Kirby so I actually got it by running:
```
PULSE_SERVER=unix:/tmp/pulse-PKdhtXMmr18n/native flatpak run com.discordapp.Discord
```
However I did notice when I run it without:
```
export XDG_RUNTIME_DIR=/run/user/$(id -u)
``` 
Added to the `/etc/profile`
I get the following errors:
```
sh: xdg-mime: command not found
```
Any idea what that may be? And what it may effect?

**Comment by kirbyfan64** [comment]
(commenting on the new issue)

**Comment by KaMiila** [comment]
Just finished a -current install and grabbed spotify with flatpak.  I was able to get audio working by creating an executable script and running spotify from that.  here is the code:

#!/bin/bash
export `xprop -root -notype PULSE_SERVER | tr -d ' '`
PULSE_SERVER=unix:/tmp/pulse-PKdhtXMmr18n/native
/usr/bin/flatpak run --branch=stable --arch=x86_64 --command=spotify --file-forwarding com.spotify.Client @@u %U @@

**Comment by IronOxidizer** [comment]
Thanks @CrashyBang and @KaMiila, your solutions worked wonderfully for my similar issue with Discord Flatpak. I'll detail my exact procedure for anyone else:

Step 1: Add the following lines to `~/.xinitrc` (before the line that starts WM)
```
mkdir -p "$HOME/.cache/xdg_runtime_dir"
export XDG_RUNTIME_DIR="$HOME/.cache/xdg_runtime_dir"
```

Step 2: Create a script (mine is ~/.local/bin/Discord) with the following lines
```
#!/bin/sh
export xprop -root -notype PULSE_SERVER | tr -d ' '
PULSE_SERVER=unix:/tmp/pulse-PKdhtXMmr18n/native
flatpak run com.discordapp.Discord
```

Step 3: Change the permission on the script to make it runnable
`chmod a+x ~/.local/bin/Discord`

Step 4: Restart/reload your WM and run the script

**Comment by smcv** [comment]
#4059 should make most (all?) of the workarounds given here unnecessary.
